### PR TITLE
perf(flattened tags) Support flattened tags on the events table

### PR DESCRIPTION
**Pull request by fpacifici**
We are adding the tags_flattened column on the events table as on the transactions table.
Here it is more complex because we have to recalculate it during replacements if the user deletes tags.

Unit test added to verify clickhouse recalculate the flattened tags correctly even when it needs to escape characters.

### CODE REVIEW

File: snuba/datasets/events_processor.py [hunk 1]
```diff
@@ -91,6 +93,30 @@ def enforce_retention(message, timestamp):
...
+    # We need to guarantee the content of the merged string is sorted otherwise we
+    # will not be able to run a LIKE operation over multiple fields at the same time.
+    # Tags are pre sorted, but it seems contexts are not, so to make this generic
+    # we ensure the invariant is respected here.
+    pairs = sorted(zip(keys, values))
+    str_pairs = [f"|{escape_field(k)}={escape_field(v)}|" for k, v in pairs]
+    # The result is going to be:
+    # |tag:val||tag:val|
+    # This gives the guarantee we will always have a delimiter on both side of the
+    # tag pair, thus we can univocally identify a tag with a LIKE expression even if
```
Comment (markstory) [inline snuba/datasets/events_processor.py]:
```suggestion
    # tag pair, thus we can unequivocally identify a tag with a LIKE expression even if
```

File: tests/test_replacer.py [hunk 2]
```diff
@@ -12,6 +12,28 @@
...
+                    tuple -> tuple.1 != %s,
+                    arrayMap((k, v) -> tuple(k, v), `tags.key`, `tags.value`)
+                )
+            )
+        ),
+        '||'
+    ),
+    '|'
+)
+"""
```
Comment (tkaemming) [inline tests/test_replacer.py]: Why isn't this sharing the definition from `snuba/replacer.py`?

File: snuba/replacer.py [hunk 3]
```diff
@@ -396,6 +424,8 @@ def process_delete_tag(message, dataset) -> Optional[Replacement]:
                 "arrayMap(x -> arrayElement(`tags.value`, x), arrayFilter(x -> x != indexOf(`tags.key`, %s), arrayEnumerate(`tags.value`)))"
                 % escape_string(tag)
             )
+        elif col.flattened == "_tags_flattened":
```
Comment (tkaemming) [inline snuba/replacer.py]: Naming is hard. 😞

File: tests/test_replacer.py [hunk 4]
```diff
@@ -108,8 +130,8 @@ def test_merge_process(self):
...
... project_id, group_id, timestamp, deleted, retention_days, platform, message, primary_hash, received, search_message, title, location, user_id, username, email, ip_address, geo_country_code, geo_region, geo_city, sdk_name, sdk_version, type, version, offset, partition, os_build, os_kernel_version, device_name, device_brand, device_locale, device_uuid, device_model_id, device_arch, device_battery_level, device_orientation, device_simulator, device_online, device_charging, level, logger, server_name, transaction, environment, `sentry:release`, `sentry:dist`, `sentry:user`, site, url, app_device, device, device_family, runtime, runtime_name, browser, browser_name, os, os_name, os_rooted, tags.key, tags.value, _tags_flattened, contexts.key, contexts.value, http_method, http_referer, exception_stacks.type, exception_stacks.value, exception_stacks.mechanism_type, exception_stacks.mechanism_handled, exception_frames.abs_path, exception_frames.filename, exception_frames.package, exception_frames.module, exception_frames.function, exception_frames.in_app, exception_frames.colno, exception_frames.lineno, exception_frames.stack_level, culprit, sdk_integrations, modules.name, modules.version",
```
Comment (tkaemming) [inline tests/test_replacer.py]: 😳

**Review by tkaemming — CHANGES_REQUESTED** [review]
Some nitpicky stuff, the redefinition of the function template is really my only big issue here.

File: tests/test_replacer.py (hunk 4 above)
Comment (fpacifici) [reply tests/test_replacer.py]: cause you did not see how it looks like in the pytest output when the two strings are not identical ...

File: snuba/datasets/events_processor.py [hunk 5]
```diff
@@ -91,6 +93,30 @@ def enforce_retention(message, timestamp):
...
+def escape_field(field: str) -> str:
+    """
+    We have ':' in our tag names. Also we may have '|'. This escapes : and \ so
+    that we can always rebuild the tags from the map. When looking for tags with LIKE
+    there should be no issue. But there may be other cases.
+    """
+    return field.translate(ESCAPE_TRANSLATION)
+
+
+def merge_nested_field(keys: Sequence[str], values: Sequence[str]) -> str:
```
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/datasets/events_processor.py]: I changed the name, transactions in general inherit some stuff from events for the reason that they are consuming the same events in the same format. We could have a common parent class between the two.

File: tests/test_replacer.py [hunk 6]
```diff
@@ -387,6 +391,84 @@ def _issue_count(total=False):
...
+                    )
+                ).encode("utf-8"),
+            ),
+            datetime.now(),
+        )
+
+        processed = self.replacer.process_message(message)
+        self.replacer.flush_batch([processed])
+
+        print(_fetch_flattened_tags())
```
Comment (tkaemming) [inline tests/test_replacer.py]: print

File: snuba/datasets/events_processor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (tkaemming) [reply snuba/datasets/events_processor.py]: Thanks, I didn't necessarily intend to suggest that the larger package structure change this needed to happen right now, but it looks good anyway. 🙂

**Review by tkaemming — APPROVED** [review]

File: tests/test_replacer.py (hunk 6 above)
Comment (fpacifici) [reply tests/test_replacer.py]: meh

File: snuba/datasets/events_processor.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fpacifici) [reply snuba/datasets/events_processor.py]: It was a quick change.

**Pull request merged**